File: workers/ruby/spec/handlers/dsl_examples/checkpoint_yield/step_handlers/checkpoint_handlers.rb
# frozen_string_literal: true

# DSL mirrors of Examples::CheckpointYield handlers.
#
# The analyzer uses batch_analyzer DSL. The worker and aggregator are
# class-based because they need direct access to Batchable instance
# methods (checkpoint_yield, batch_worker_success, handle_no_op_worker,
# detect_aggregation_scenario, etc.).

include TaskerCore::StepHandler::Functional # rubocop:disable Style/MixinUsage

# DSL mirror of Examples::CheckpointYield::AnalyzerHandler
AnalyzeItemsDslHandler = batch_analyzer 'checkpoint_yield_dsl.step_handlers.analyze_items',
                                        worker_template: 'checkpoint_yield_batch_dsl',
                                        inputs: [:total_items] do |total_items:, context: nil| # rubocop:disable Lint/UnusedBlockArgument
  total = total_items || 100

  if total <= 0
    next TaskerCore::Types::StepHandlerCallResult.success(
      result: { 'batch_processing_outcome' => 'no_batches', 'reason' => 'no_items_to_process' }
    )
  end

  TaskerCore::StepHandler::Functional::BatchConfig.new(
    total_items: total,
    batch_size: total,
    metadata: {
      'test_type' => 'checkpoint_yield',
      'items_per_batch' => total
    }
  )
end

# Class-based DSL mirror of Examples::CheckpointYield::WorkerHandler
class CheckpointYieldBatchDslHandler < TaskerCore::StepHandler::Batchable
  def handler_name
    'checkpoint_yield_dsl.step_handlers.checkpoint_yield_batch'
  end

  def call(context)
    batch_ctx = get_batch_context(context)
    return batch_worker_failure('No batch inputs found') unless batch_ctx

    no_op_result = handle_no_op_worker(batch_ctx)
    return no_op_result if no_op_result

    items_per_checkpoint = context.get_input_or(
      'items_per_checkpoint', context.get_config('items_per_checkpoint') || 25
    )
    fail_after_items = context.get_input('fail_after_items')
    fail_on_attempt = context.get_input_or('fail_on_attempt', 1)
    permanent_failure = context.get_input_or('permanent_failure', false)

    if batch_ctx.has_checkpoint?
      start_cursor = batch_ctx.checkpoint_cursor
      accumulated = batch_ctx.accumulated_results || { 'running_total' => 0, 'item_ids' => [] }
      total_processed = batch_ctx.checkpoint_items_processed
    else
      start_cursor = batch_ctx.start_cursor
      accumulated = { 'running_total' => 0, 'item_ids' => [] }
      total_processed = 0
    end

    end_cursor = batch_ctx.end_cursor
    current_attempt = context.retry_count + 1
    current_cursor = start_cursor
    items_in_chunk = 0

    while current_cursor < end_cursor
      if fail_after_items && total_processed >= fail_after_items && current_attempt == fail_on_attempt
        return inject_failure(total_processed, current_cursor, permanent_failure)
      end

      accumulated['running_total'] += (current_cursor + 1)
      accumulated['item_ids'] << "item_#{current_cursor.to_s.rjust(4, '0')}"

      current_cursor += 1
      items_in_chunk += 1
      total_processed += 1

      next unless items_in_chunk >= items_per_checkpoint && current_cursor < end_cursor

      return checkpoint_yield(
        cursor: current_cursor,
        items_processed: total_processed,
        accumulated_results: accumulated
      )
    end

    batch_worker_success(
      items_processed: total_processed,
      items_succeeded: total_processed,
      items_failed: 0,
      results: accumulated.merge(
        'final_cursor' => current_cursor,
        'checkpoints_used' => total_processed / items_per_checkpoint
      )
    )
  end

  private

  def inject_failure(items_processed, cursor, permanent)
    error_type = permanent ? 'PermanentError' : 'RetryableError'
    failure_type = permanent ? 'permanent' : 'transient'

    TaskerCore::Types::StepHandlerCallResult.error(
      message: "Injected #{failure_type} failure after #{items_processed} items",
      error_type: error_type,
      retryable: !permanent,
      metadata: {
        items_processed: items_processed,
        cursor_at_failure: cursor,
        failure_type: failure_type
      }
    )
  end
end

# Class-based DSL mirror of Examples::CheckpointYield::AggregatorHandler
class AggregateResultsDslHandler < TaskerCore::StepHandler::Batchable
  def handler_name
    'checkpoint_yield_dsl.step_handlers.aggregate_results'
  end

  def call(context)
    scenario = detect_aggregation_scenario(
      context.dependency_results,
      'analyze_items_dsl',
      'checkpoint_yield_batch_dsl_'
    )

    if scenario.no_batches?
      return no_batches_aggregation_result(
        'total_processed' => 0,
        'running_total' => 0,
        'test_passed' => true,
        'scenario' => 'no_batches'
      )
    end

    total_processed = 0
    running_total = 0
    all_item_ids = []
    checkpoints_used = 0

    scenario.batch_results.each_value do |result|
      next unless result

      total_processed += result['items_processed'] || 0
      batch_data = result['results'] || {}
      running_total += batch_data['running_total'] || 0
      all_item_ids.concat(batch_data['item_ids'] || [])
      checkpoints_used += batch_data['checkpoints_used'] || 0
    end

    TaskerCore::Types::StepHandlerCallResult.success(
      result: {
        'total_processed' => total_processed,
        'running_total' => running_total,
        'item_count' => all_item_ids.length,
        'checkpoints_used' => checkpoints_used,
        'worker_count' => scenario.worker_count,
        'test_passed' => true,
        'scenario' => 'with_batches'
      }
    )
  end
end

File: crates/workers/typescript/tests/handlers/dsl_examples/checkpoint_yield/step_handlers/checkpoint-handlers.ts
/**
 * Checkpoint Yield DSL Step Handlers (TAS-294).
 *
 * The analyzer uses defineBatchAnalyzer factory. The worker and aggregator
 * are class-based because they need Batchable instance methods:
 * checkpointYield(), handleNoOpWorker(), getBatchWorkerInputs(),
 * batchWorkerSuccess(), etc.
 */

import { StepHandler } from '../../../../../src/handler/base.js';
import { BatchableStepHandler } from '../../../../../src/handler/batchable.js';
import { defineBatchAnalyzer } from '../../../../../src/handler/functional.js';
import type { StepContext } from '../../../../../src/types/step-context.js';
import type { StepHandlerResult } from '../../../../../src/types/step-handler-result.js';

const DEFAULT_TOTAL_ITEMS = 100;
const DEFAULT_ITEMS_PER_CHECKPOINT = 25;

/**
 * TAS-125: Checkpoint Yield Analyzer Handler (DSL).
 */
export const CheckpointYieldAnalyzerDslHandler = defineBatchAnalyzer(
  'checkpoint_yield_dsl.step_handlers.CheckpointYieldAnalyzerDslHandler',
  {
    workerTemplate: 'checkpoint_yield_batch_dsl_ts',
  },
  async ({ context }) => {
    const totalItems = context.getInputOr(
      'total_items',
      context.getConfig<number>('total_items') ?? DEFAULT_TOTAL_ITEMS
    );

    if (totalItems <= 0) {
      return { totalItems: 0, batchSize: 1 };
    }

    // Single batch for checkpoint test
    return {
      totalItems,
      batchSize: totalItems,
      metadata: {
        test_type: 'checkpoint_yield',
        total_items: totalItems,
        analyzed_at: new Date().toISOString(),
      },
    };
  }
);

/** Configuration for checkpoint yield worker */
interface WorkerConfig {
  itemsPerCheckpoint: number;
  failAfterItems: number | undefined;
  failOnAttempt: number;
  permanentFailure: boolean;
}

/** Processing state for checkpoint yield worker */
interface ProcessingState {
  startCursor: number;
  accumulated: { running_total: number; item_ids: string[] };
  totalProcessed: number;
  currentAttempt: number;
}

/**
 * TAS-125: Checkpoint Yield Worker Handler (DSL).
 *
 * Class-based because it needs checkpointYield(), handleNoOpWorker(),
 * and getBatchWorkerInputs() from BatchableStepHandler.
 */
export class CheckpointYieldWorkerDslHandler extends BatchableStepHandler {
  static handlerName = 'checkpoint_yield_dsl.step_handlers.CheckpointYieldWorkerDslHandler';
  static handlerVersion = '1.0.0';

  async call(context: StepContext): Promise<StepHandlerResult> {
    // Check for no-op placeholder
    const noOpResult = this.handleNoOpWorker(context);
    if (noOpResult) {
      return noOpResult;
    }

    // Get batch worker inputs
    const batchInputs = this.getBatchWorkerInputs(context);
    const cursor = batchInputs?.cursor;

    if (!cursor) {
      return this.failure('No batch inputs found', 'batch_error', false);
    }

    const config = this.getWorkerConfig(context);
    const state = this.getProcessingState(context, cursor.start_cursor);
    const endCursor = cursor.end_cursor;

    return this.processItems(config, state, endCursor);
  }

  private getWorkerConfig(context: StepContext): WorkerConfig {
    return {
      itemsPerCheckpoint: context.getInputOr(
        'items_per_checkpoint',
        context.getConfig<number>('items_per_checkpoint') ?? DEFAULT_ITEMS_PER_CHECKPOINT
      ),
      failAfterItems: context.getInput<number>('fail_after_items') ?? undefined,
      failOnAttempt: context.getInputOr('fail_on_attempt', 1),
      permanentFailure: context.getInputOr('permanent_failure', false),
    };
  }

  private getProcessingState(context: StepContext, defaultStartCursor: number): ProcessingState {
    const currentAttempt = context.retryCount + 1;

    if (context.hasCheckpoint()) {
      return {
        startCursor: (context.checkpointCursor as number) ?? defaultStartCursor,
        accumulated: (context.accumulatedResults as ProcessingState['accumulated']) ?? {
          running_total: 0,
          item_ids: [],
        },
        totalProcessed: context.checkpointItemsProcessed,
        currentAttempt,
      };
    }

    return {
      startCursor: defaultStartCursor,
      accumulated: { running_total: 0, item_ids: [] },
      totalProcessed: 0,
      currentAttempt,
    };
  }

  private processItems(
    config: WorkerConfig,
    state: ProcessingState,
    endCursor: number
  ): StepHandlerResult {
    let currentCursor = state.startCursor;
    let itemsInChunk = 0;
    let totalProcessed = state.totalProcessed;
    const accumulated = state.accumulated;

    while (currentCursor < endCursor) {
      // Failure injection
      if (this.shouldInjectFailure(config, totalProcessed, state.currentAttempt)) {
        return this.injectFailure(totalProcessed, currentCursor, config.permanentFailure);
      }

      // Process one item
      accumulated.running_total += currentCursor + 1;
      accumulated.item_ids.push(`item_${String(currentCursor).padStart(4, '0')}`);

      currentCursor += 1;
      itemsInChunk += 1;
      totalProcessed += 1;

      // Yield checkpoint at interval
      if (itemsInChunk >= config.itemsPerCheckpoint && currentCursor < endCursor) {
        return this.checkpointYield(currentCursor, totalProcessed, accumulated);
      }
    }

    // All items processed
    return this.success({
      items_processed: totalProcessed,
      items_succeeded: totalProcessed,
      items_failed: 0,
      batch_metadata: {
        ...accumulated,
        final_cursor: currentCursor,
        checkpoints_used: Math.floor(totalProcessed / config.itemsPerCheckpoint),
      },
    });
  }

  private shouldInjectFailure(
    config: WorkerConfig,
    totalProcessed: number,
    currentAttempt: number
  ): boolean {
    return (
      config.failAfterItems !== undefined &&
      totalProcessed >= config.failAfterItems &&
      currentAttempt === config.failOnAttempt
    );
  }

  private injectFailure(
    itemsProcessed: number,
    cursor: number,
    permanent: boolean
  ): StepHandlerResult {
    const errorType = permanent ? 'PermanentError' : 'RetryableError';
    const failureType = permanent ? 'permanent' : 'transient';
    const message = `Injected ${failureType} failure after ${itemsProcessed} items`;

    return this.failure(message, errorType, !permanent, {
      items_processed: itemsProcessed,
      cursor_at_failure: cursor,
      failure_type: failureType,
    });
  }
}

/**
 * TAS-125: Checkpoint Yield Aggregator Handler (DSL).
 */
export class CheckpointYieldAggregatorDslHandler extends StepHandler {
  static handlerName = 'checkpoint_yield_dsl.step_handlers.CheckpointYieldAggregatorDslHandler';
  static handlerVersion = '1.0.0';

  async call(context: StepContext): Promise<StepHandlerResult> {
    const batchResults = context.getAllDependencyResults(
      'checkpoint_yield_batch_dsl_ts'
    ) as Array<Record<string, unknown> | null>;

    // Handle no batches scenario
    const analyzeResult = context.getDependencyResult('analyze_items_dsl_ts') as Record<
      string,
      unknown
    > | null;
    const outcome = analyzeResult?.batch_processing_outcome as Record<string, unknown> | undefined;

    if (outcome?.type === 'no_batches') {
      return this.success({
        total_processed: 0,
        running_total: 0,
        test_passed: true,
        scenario: 'no_batches',
      });
    }

    if (!batchResults || batchResults.length === 0) {
      return this.failure('No batch worker results to aggregate', 'aggregation_error', false);
    }

    let totalProcessed = 0;
    let runningTotal = 0;
    const allItemIds: string[] = [];
    let checkpointsUsed = 0;

    for (const result of batchResults) {
      if (!result) continue;

      totalProcessed += (result.items_processed as number) ?? 0;
      const batchMetadata = result.batch_metadata as Record<string, unknown> | undefined;
      if (batchMetadata) {
        runningTotal += (batchMetadata.running_total as number) ?? 0;
        const itemIds = batchMetadata.item_ids as string[] | undefined;
        if (itemIds) {
          allItemIds.push(...itemIds);
        }
        checkpointsUsed += (batchMetadata.checkpoints_used as number) ?? 0;
      }
    }

    return this.success({
      total_processed: totalProcessed,
      running_total: runningTotal,
      item_count: allItemIds.length,
      checkpoints_used: checkpointsUsed,
      worker_count: batchResults.length,
      test_passed: true,
      scenario: 'with_batches',
    });
  }
}
